mozci/data/sources/treeherder/__init__.py

```diff
@@ -3,24 +3,71 @@
 from collections import defaultdict
 from functools import lru_cache
 
+from adr.util.memoize import memoized_property
 from loguru import logger
 
 from mozci.data.base import DataSource
 from mozci.errors import ContractNotFilled
+from mozci.util.req import get_session
 
 try:
     from treeherder.model.models import Job
 except ImportError:
     Job = None
 
 
-class TreeherderSource(DataSource):
-    """Uses ORM to query Treeherder."""
+class BaseTreeherderSource(DataSource):
+    pass
 
-    name = "treeherder"
+
+class TreeherderClientSource(BaseTreeherderSource):
+    """Uses the public API to query Treeherder."""
+
+    name = "treeherder_client"
+    supported_contracts = ("push_tasks_classifications",)
+    base_url = "https://treeherder.mozilla.org/api"
+
+    @memoized_property
+    def session(self):
+        session = get_session()
+        session.headers = {"User-Agent": "mozci"}
+        return session
+
+    def _run_query(self, query):
+        query = query.strip("/")
+        url = f"{self.base_url}/{query}"
+        if "?" in query:
+            url += "&format=json"
+        else:
+            url += "/?format=json"
+
+        response = self.session.get(url)
+        response.raise_for_status()
+        return response.json()
+
+    def run_push_tasks_classifications(self, branch, rev):
+        data = self._run_query(f"/project/{branch}/note/push_notes/?revision={rev}")
+
+        classifications = {}
+        for item in data:
+            job = item["job"]
+            classifications[job["task_id"]] = {
+                "classification": item["failure_classification_name"]
+            }
+            if item["text"]:
+                classifications[job["task_id"]]["failure_classification_name"] = item[
```

Comment:
This should be `classification_note`, right?
How come voluptous is not catching the schema mismatch?

Reply:
Good catch. Turns out the push I randomly selected to test on happened to not have any classification notes on it. I updated the test to make sure we validate the output against the contract.